chore(node): add support for webpack

boneskull opened the pull request:
This is here because I didn't want to pile more stuff on #1543.

At some point, @naugtur had requested we rename `--root` to `--project-root` so that's happening here.  

**UPDATE Oct 31 2025**: There are many other changes which integrate with endojs/endo#2988.  Thus, CI has no hope of passing until that PR is merged, which is why this is DRAFT.

naugtur commented on packages/node/src/cli.js:
```diff
@@ -420,48 +412,43 @@ const main = async (args = hideBin(process.argv)) => {
-          })
         }
 
         stripProcessArgv(
```
stripProcessArgv is incorrect. It takes argv['--'] from yargs and splices it in front of arguments that follow the first `--` , so it should be duplicating it if I'm reading it correctly.

seems like we want to replace process.argv with the items array, not splice it in.

Added a failing test for that.

boneskull replied:
- [x] do this

naugtur commented on packages/node/src/policy-util.js:
Some of the stuff here would be good to have in core so that it benefits all implementations

naugtur commented on packages/node/src/exec/exec-compartment-class.js:
(must) reuse the one from core

boneskull replied:
You just added this, right?

naugtur replied:
The function was there for a long time, I added a fix to it for one particularly unexpected corner case where constructors have a prototype chain.

naugtur replied:
fixed

naugtur commented on packages/node/src/exec/default-attenuator.js:
(must) disallow the deprecated type of scuttling definition. If you need a boolean arg in cli, convert it to the new shape before it reaches any further. 
Use the type that doesn't allow it to be boolean. And finally, remove this block.


```suggestion
```

boneskull replied:
- [ ] do this

naugtur commented on packages/node/src/compartment/node-compartment-map.js:
```diff
@@ -0,0 +1,180 @@
+     * Adds any missing dependencies from `policyOverride` to the list of
+     * dependencies for a package
+     */
+    packageDependenciesHook: ({ canonicalName, dependencies }) => {
```
(could) Consider making this hook's existence conditional on policyOverride as an optimization

naugtur commented on packages/node/src/exec/load.js:
```diff
@@ -7,15 +7,18 @@
- * @import {ReadNowPowers, LoadLocationOptions} from '@endo/compartment-mapper'
+ * @import {ImportLocationOptions, SyncImportLocationOptions, CanonicalName, PackageCompartmentMapDescriptor} from '@endo/compartment-mapper'
  * @import {ApplicationLoader, ExecuteOptions} from '../types.js'
+ * @import {PackageJson} from 'type-fest'
```
unused import?

naugtur commented on packages/node/src/report.js:
```diff
@@ -0,0 +1,224 @@
+ * @returns {void}
+ * @internal
+ */
+export const reportInvalidCanonicalNames = (
```
(not in this PR) Having this for webpack as well would be great. I'd like to migrate this to core or somewhere shared.

naugtur commented on packages/node/src/report.js:
```diff
@@ -0,0 +1,224 @@
+    const suggestions = new Map()
+    for (const { name: invalidName } of invalidCanonicalNames) {
+      const invalidNameParts = invalidName.split('>')
+      const invalidPackageName = invalidNameParts[invalidNameParts.length - 1]
```
(nit) the package name is not actually invalid. `misplacedPackageName` maybe?

naugtur commented on packages/node/src/report.js:
```diff
@@ -0,0 +1,224 @@
+    })
+    .filter(({ source }) => source !== undefined)
+
+  // if we have any invalid overrides, we will search through the canonical names from the compartment map and make suggestions for the user to fix them
```
clever!

naugtur commented on packages/node/src/exec/load.js:
```diff
@@ -29,36 +32,73 @@ import { toEndoURL } from '../util.js'
   }
 
-  const { import: importApp, sha512 } = await loadLocation(
+  const { import: importApp, sha512 } = await loadFromMap(
```
(not for this PR)
We never look at the contents of `packageCompartmentMap`

mapNoceModules might as well return an object with loadFrom map method. 

In a cleaner fashion:

```js
// full chain
makeApp({ subset of loadLocation options })
  .mapNodeModules({existing options})
  .load({same as loadFromMap})
  .import({the existing one})

// all-default mapping
makeApp({ subset of loadLocation options })
  .load({same as loadFromMap})
  .import({the existing one})

// simplest
makeApp({ subset of loadLocation options })
  .import({the existing one})
```

There's room to return more, so mapNodeModules could still return a reference to compartmentMap, but the user is not responsible for passing it around to load.

boneskull replied:
IDK man, I'd rather chew broken glass than implement a fluent API.  And I know I'm  not as offended by the current API as yourself (though I agree it could be streamlined, I could live in bliss never touching it again).  But we should brainstorm a bit more, I think.

naugtur commented on packages/node/src/types.ts:
```diff
@@ -275,27 +282,52 @@ export type CompartmentMapToDebugPolicyOptions = SetFieldType<
+    WithTrustRoot,
+    WithScuttleGlobalThis,
+    WithReadFile,
+    WithProjectRoot,
```
(uncertain) 
This contains something like 40 options. There's no good way to read about them all in one place. 
I think generatePolicy options got out of hand. 

Maybe splitting them into groups would help. policy paths, root and scuttling separately from the bag of tools like readPowers and isBuiltin separately from load options etc.

boneskull replied:
I'll see if Claude can help me with this one 😜 

- [x] do this

boneskull replied:
Well, wait a minute. There is a place to view most of them all at once:

<img width="966" height="596" alt="image" src="https://github.com/user-attachments/assets/b4f26dc1-0dd0-4248-8a26-09831c5a4386" />

But we can't view ALL of them at once, because there's just a ton of them (like even 50), and the vast majority are passed thru to Endo's `loadFromMap()`.

- I think we can make this a little less awful by _nesting_ the Endo-specific options.
- I can also try to logically group these options by identifying which always appear together.

boneskull replied:
Nesting the options is proving difficult.  Will continue plugging at it though, since I think it makes the most sense.  This would add `loadForMapOptions` & `mapNodeModulesOptions` fields to `ExecuteOptions` and a `mapNodeModulesOptions` field to `GenerateOptions`.

boneskull replied:
I did some consolidation but I think I'd like to work on the options further in a subsequent PR.

naugtur commented on packages/node/src/types.ts:
```diff
@@ -275,27 +282,52 @@ export type CompartmentMapToDebugPolicyOptions = SetFieldType<
+  [
+    WithReadPowers,
+    WithIsBuiltin,
+    WithLoadForMapOptions,
```
Is his a misspelling of `LoadFromMapOptions`?

boneskull replied:
No, it's a type which takes most of endo's `LoadForMapOptions` and throws out the stuff we're going to explicitly provide.

naugtur commented on packages/node/src/types.ts:
```diff
@@ -167,16 +195,25 @@ export interface WithReadPowers {
   enabled?: boolean
   exceptions?: string[]
   scuttlerName?: string
 }
```
(must) this is wrong. Import the type from core. exception are `Array<string|RegExp>`

boneskull replied:
Ah, I think this was leftover from before the type existed.

- [x] do this

naugtur commented on packages/node/src/policy-gen/generate.js:
```diff
@@ -5,306 +5,120 @@
-    const { debugInfo: _, ...corePolicy } = /** @type {LavaMoatPolicyDebug} */ (
-      debugPolicy
+  } else if (rawPolicyOverridePath) {
+    log.warning(
```
(should) I'd make this an error if input has both

naugtur commented on packages/node/src/policy-gen/generate.js:
```diff
@@ -5,306 +5,120 @@
     await writePolicy(policyPath, policy, { fs: writableFs })
-    log.info(`Wrote policy to ${hrPath(policyPath)}`)
+    log.info(
+      `${chalk.greenBright('✓')} ${chalk.bold('Wrote policy')} ${chalk.white('to')} ${hrPath(policyPath)}`
```
(should) Eliminate the dependency on chalk. We should limit our prod dependencies to necessities. Colorful output is not worth it. It'll run in CI half the time anyway.

boneskull replied:
I'm gonna push back on the colors, but we can just vendor some ANSI codes.

naugtur replied:
I'm indifferent about colors, personally prefer ascii art or shapes. Just saying it's not worth having a dependency for it if that dependency introduces risk of malware.
vendoring ANSI codes as functions with template strings (so a subset of chalk I guess, but maybe lacking some compatibility fanciness) sounds better.
(I know we can't get rid of chalk from dependencies entirely yet, but in preparation for that)

boneskull replied:
Just to note here, a switch from `loggerr` to `@lavamoat/vog` will eliminate the dependency on `chalk`.

naugtur commented on packages/node/src/policy-gen/load-for-policy.js:
```diff
@@ -0,0 +1,463 @@
+          }
+          if (violations) {
+            const currentViolations = violationsForPackage.get(canonicalName)
+            if (currentViolations) {
```
(should) this function mixes come very distinct responsibilities. It understands messages from workers, displaying progress but also structure of policy and types of violations. 
If we add a new violation, I don't want to go and update this function. I'd rather generalize the aggregation and/or make it happen elsewhere.

boneskull replied:
By "elsewhere" you mean in the parent process somewhere?  Or "elsewhere" in some separate function?  I agree it could be split up.

naugtur replied:
I mean splitting the code into separate functions/files to keep the logic and the workers code not entangled (as in: if you wanted to rip out workers, what structure of the code would make it the easiest to do without risking introducing bugs in logic?)

boneskull replied:
OK, I've split this up into like four functions and moved the reporting business into `report.js`.

naugtur commented on packages/node/src/policy-gen/load-for-policy.js:
```diff
@@ -0,0 +1,463 @@
+   *
+   * @type {Map<CanonicalName, PackagePolicy>}
+   */
+  const packagesForPackage = new Map()
```
(could) rename to `connectionsForPackage` ?

boneskull replied:
`packagePoliciesMap` seemed more appropriate to me, so there it is.

naugtur commented on packages/node/src/report.js:
```diff
@@ -0,0 +1,224 @@
+        source: keypath || `unknown location for "${canonicalName}"`,
+      }
+    })
+    .filter(({ source }) => source !== undefined)
```
source can't be undefined if it has a fallback to a string.

naugtur commented on packages/node/src/policy-gen/load-for-policy.js:
```diff
@@ -0,0 +1,463 @@
+        if ('location' in moduleSource) {
+          return inspectModuleSource(moduleSource, canonicalName)
+        } else if ('error' in moduleSource) {
+          throw new GenerationError(
```
Do you know what circumstances this could happen under?

boneskull replied:
It can't; I think I missed this after some refactoring.

naugtur commented on packages/node/src/policy-gen/inspector.js:
```diff
@@ -0,0 +1,380 @@
+ *
+ * This function:
+ *
+ * 1. Decodes the bytes into a string using `TextDecoder` (utf-8 encoding).
```
that's not true. looks like the comment is stale

naugtur commented on packages/node/src/policy-gen/inspector.js:
```diff
@@ -0,0 +1,380 @@
+ * @returns {StructuredViolationsResult | null}
+ */
+const inspectViolations = (ast, id) => {
+  const compatWarnings = inspectSesCompat(/** @type {any} */ (ast))
```
the type casting here should not be needed. This being necessary suggests we might be depending on a different version of babel

boneskull replied:
`inspectSesCompat` mutates the `error` property of the AST in such a way that the AST is no longer structurally equivalent to a `ParseResult`.  This is bad, but that's what it does.  There's likely something in `lavamoat-node` that's relying on this behavior, so I'm loathe to try to fix it.

naugtur replied:
If you're sure it's not just types getting mixed up, let's add an issue for fixing it and move on.

boneskull replied:
Created https://github.com/LavaMoat/LavaMoat/issues/1874

naugtur commented on packages/node/src/policy-gen/inspector.js:
```diff
@@ -0,0 +1,380 @@
+ *    - Splits the key on the first `.` (this array must have length >= 2)
+ *    - Creates a new entry choosing the head element of the array as the new key and
+ *         the value as `true`
+ *    - Deletes the original entry
```
(must) 
Why would htere be a nested property marked as writable? 

`"someproperty":true` means you get a reference to someproperty and `someproperty.nested` is writeable without asking (this gets complicated with global getters and functions)

`"someproperty": "write"` means you can assign to it and it will propagate to other compartments.

AFAIR tofu should not return nested properties for writable. If it does, collapsing them to the top property doesn't seem to make sense. 

We should look at an example of this and fugure out why this is here.

boneskull replied:
> Why would htere be a nested property marked as writable?

Because this is what tofu does when it finds a global nested property which is written to.  I was under the impression—at least historically—that tofu was correct and we simply didn't support this in the kernel; I'm pretty sure we have code elsewhere that works around the problem.  I don't think I should be changing this behavior in tofu as part of the PR, though.

naugtur replied:
We need to discuss and explore, this TOFU behavior was earlier masked by policy merging collapsing nested fields (sometimes leading to incorrect results) 

```
"sth": true,
"sth.more": write
```
Would end up collapsed to just 
```
"sth": true
```

boneskull replied:
Right, we did that as a crappy alternative to actually implementing the behavior in the kernel so that it'd work.  ¯\\\_(ツ)\_/¯

naugtur commented on packages/node/src/policy-gen/load-for-policy.js:
```diff
@@ -0,0 +1,463 @@
+   * @param {Omit<
+   *   LocalModuleSource,
+   *   'location' | 'sourceLocation' | 'sourceDirname' | 'parser' | 'record'
+   * > & { language: Language; location: FileUrlString }} moduleSource
```
Is this an artifact of waiting for latest Endo, or does the type really differ frmo the value we're getting here? I f so, why is it rewired here and not defined somewhere as a new type for our internal use? I'd like to get the origin story of what's going on here.

boneskull replied:
It's the `ModuleSourceForModuleSourceHook` type, which didn't exist at one point.  Now it does and I can use it.

naugtur commented on packages/node/src/policy-gen/generate.js:
```diff
@@ -5,306 +5,120 @@
 
 /**
  * Generates a LavaMoat policy or debug policy from a given entry point using
  * `@endo/compartment-mapper`
```
(nit) I don't think the description is true. we no longer do debug policy

naugtur commented on packages/node/src/error.js:
```diff
@@ -92,9 +84,12 @@ export const InvalidPolicyError = createLavaMoatError(
+  ErrorCodes.NoBinScript,
   class extends Error {}
 )
 
```
seems the file has not been formatted

naugtur commented on packages/node/src/policy-util.js:
```diff
@@ -11,21 +11,38 @@
 import { jsonStringifySortedPolicy, mergePolicy } from 'lavamoat-core'
```
`TODO: Capabilities are inconsistent`
Care to elaborate?

boneskull replied:
Some of the functions accept readPowers "capabilities", and others don't. I'm pretty sure I'm using them somewhere in the tests to read data from JSON fixtures, at any rate.  Shall I elaborate more in the comment?

naugtur replied:
Since it's a TODO, elaborating a little in the comment so that the intention is clear to someone else would make it easier to address the todo by someone who's not in the state of mind to know what's to be done from just looking. (eg. future self :P )

naugtur commented on packages/node/src/policy-util.js:
```diff
@@ -11,21 +11,38 @@
 import { jsonStringifySortedPolicy, mergePolicy } from 'lavamoat-core'
 import nodeFs from 'node:fs'
 import nodePath from 'node:path'
+import { toKeypath } from 'to-keypath'
```
nice one, but let's vendor it. No need to have a prod dependency and the risk is always there.

naugtur commented on packages/node/src/compartment/import-hook.js:
we must check with isBuiltin here and error out otherwise, unless we can guarantee isBuiltin is called in all paths leading to this (which I don't think is the case)

boneskull replied:
So the thing is builtins go thru the exit module, but so do things like JSON files and native modules.  We can do a naive check of the filename at this point (`.json`/`.node`) as well as checking `isBuiltin()`, but that's about all we can do.

I note that Endo actually calls the exit module import hook _before_ checking policy (and I still don't really understand why) which means if we add this check we're going to hit it even before we report invalid canonical names, and invalid canonical names will _always_ be considered exit modules.

https://github.com/endojs/endo/blob/9463241e6014366ba9956cf01bd4299119a78b9d/packages/compartment-mapper/src/import-hook.js#L813-827

So I think we need to discuss further and address this post-merge.

naugtur replied:
AFAIR the problem was we need the error that it's missing reported before it complain about policy. The memory is faded, but I remember eing frustrated that I have to call it before checking the policy for some specific reason I couldn't work around at the time.

we have parsers defined for endo for .json and .node - why are they going through exit module import hook?

We definitely need to talk about it more.

boneskull replied:
Hmm, right. We do have parsers. 

I think I want to experiment with that code in Endo. Because IMO it's _really_ bad behavior the way it works now.

naugtur reviewed (approved):
renewing the approval

boneskull committed:
chore: fix lint-staged config to avoid unprettifiable files

The previous globs were a heroic effort, but ultimately flawed.  This ensures we do not prettify _anything_ in a `node_modules` dir, which includes test fixtures.

boneskull committed:
chore(vscode): do not search generated types and distfiles

boneskull committed:
chore(ci): ensure setup runs before linting happens

tseslint has issues with types referencing anything in `@lavamoat/types` if its distfiles aren't built.

boneskull committed:
chore(browserify): skip builtin scenario

This does not make sense in a browserify context.

boneskull committed:
feat(core): export MODULE_REFS constant; fix strict mode in function wrapper

- Add some internal types

boneskull committed:
fix(deps): upgrade ses, @types/node, endo

- Adds `c8`
- Adds `@babel/parser`, `lavamoat-tofu`, and `@lavamoat/types` to `@lavamoat/node`

boneskull committed:
chore(tofu): fix some internal types

boneskull committed:
feat(types): add `include` property to LavaMoatPolicy

- Other slight type fixes

boneskull committed:
feat(node)!: major compatibility changes & parallel module inspection

BREAKING CHANGES: Rebuild `@lavamoat/node`'s policy generation and runtime execution fully top of `@endo/compartment-mapper` v2's new hooks, replacing the previous `LavamoatModuleRecord`-based module loading and linking. `NoExecutableError` is now `NoBinScriptError`. Tightened exports. Support for "debug" policies dropped.

## Policy generation & conversion

Policy is now generated from `@endo/compartment-mapper`'s compartment
map and then converted to LavaMoat's format. Several bugs were found
and fixed during this conversion:

- Redundant builtins (e.g., both `fs` and `fs.readFile` appearing) are
  now de-duplicated
- Redundant globals entries are eliminated (alternatively `mergePolicy`
  in core could handle this, but doing it at generation time is simpler)
- `$root$` is now explicitly allowed to be referenced in LavaMoat policy
- Policy object validation was tightened from its initial implementation
- A warning is now emitted when a non-builtin module is loaded as if it
  were a builtin
- Module inspections (AST parsing) now happens in a pool of worker
  threads for efficiency

## Trusted roots & global attenuation

When a trusted root exists, its `Compartment`'s `globalThis` must be
created _before_ any dependent compartment, since dependents derive
their globals from it. `@endo/compartment-mapper`'s linker does not
guarantee attenuation order, so a queue-based deferral mechanism was
introduced: attenuations are enqueued until the root compartment is
encountered, then the queue is synchronously drained.

Writable globals (`write` in policy) were also broken and are now fixed.

## Evasive transforms

Adopted new functionality in `@endo/evasive-transform` instead of a local
implementation. Tracked several upstream API changes across
endojs/endo#2988, endojs/endo#2915, and endojs/endo#3111.

## CLI: `--dev` replaced by `--prod-only`

Dev dependencies are now included by default, which matches what most
users expect. The old `--dev` opt-in flag is replaced by `--prod-only`
to exclude them. Internally, all functions accept `prodOnly` and the
bit is flipped to `dev: !prodOnly` only at the two Endo API boundary
call sites.

`--root` is now `--project-root`, but `--root` will still work.

Fixed problems with double-dash (`--`) arg parsing.

## Other notable changes

- `include` policy field support (force packages into generated policy)
- `WritePolicyError` for better error reporting during policy writes
- Exit module warnings are aggregated to avoid clobbering progress
  indicators
- Spurious "unknown canonical name" warnings suppressed for untrusted
  entrypoints
- Better heuristic for guessing the policy override filename
- Pluralized user-facing messages

## Additional Test coverage

- E2E: `--dev`/`--prod-only`, `$root$` in policy, missing runtime
  packages, untrusted entrypoints, exit module behavior during policy
  gen, phantom packages (exist on disk but not in `package.json`)
- Unit: CLI arg reassembly, evasive transforms, policy generation,
  worker pool

## Internal Tooling

- `LavaMoatError`-implementing constructors can use the new generic
  `LavaMoatErrorOptions`
- `Loggerr` re-exported from `log.js` module for referencing loglevels
- Expose hidden/experimental `--trust-root` for `run` command to force
  trusting the root `Compartment`. Useful for testing.
